cpp/src/groupby/groupby.cu

```diff
@@ -113,11 +138,75 @@ groupby::aggregate(std::vector<aggregation_request> const& requests,
   verify_valid_requests(requests);
 
   if (_keys.num_rows() == 0) {
-    std::make_pair(empty_like(_keys), empty_results(requests));
+    std::make_pair(empty_like(_keys), 
+                   templated_empty_results(requests, 
+                                           [](std::unique_ptr<aggregation> const& agg)
+                                           {return agg->kind;}));
   }
 
   return dispatch_aggregation(requests, 0, mr);
 }
+
+// Get the sort helper object
+detail::sort::sort_groupby_helper& groupby::helper() {
+  if (_helper)
+    return *_helper;
+  _helper = std::make_unique<detail::sort::sort_groupby_helper>(
+    _keys, _ignore_null_keys, _keys_are_sorted);
+  return *_helper;
+};
+
+std::pair<std::unique_ptr<table>, std::vector<aggregation_result>> groupby::aggregate(
+    std::vector<window_aggregation_request> const& requests,
+    rmm::mr::device_memory_resource* mr) {
+
+  CUDF_EXPECTS(std::all_of(requests.begin(), requests.end(),
+                           [this](auto const& request) {
+                             return request.values.size() == _keys.num_rows();
+                           }),
+               "Size mismatch between request values and groupby keys.");
+
+  CUDF_EXPECTS(this->_keys_are_sorted, 
```

Comment:
This doesn't have to be. It's pretty easy to sort the input, unless that's not a priority at all.

Reply:
Sorting isn't currently a priority. At least from Spark's perspective, this data will come in pre-sorted. I figured we could add in support for sorting later on, if required.